### PR TITLE
Redesign Friends Component, Create Group Invitations Tab, Create Group & DM

### PR DESCRIPTION
**Pull request by bhunt02**
# Big Ol' Pull Request of Miscellaneous Issues because we're 1 week off the due date and I can't be bothered to split these up

Here we go

## Additions
* Navlayout
  * Button in Group Chat dropdown "Create Group" that triggers a callback to open up a popup to create group chats (more on this later)
  * Button in DMs dropdown that triggers a callback to open up a popup to open a DM (just from friend's list; set-up with same button from friend list on (renamed) "home" tab and user profiles "send message" that either requests to create the DM (if it does not exist) or uses a callback that opens the chat
    * There's no reason for some overcomplicated popup to 'create' a DM, since it's just between friends.
* Create Group Popup
  * It's what it sounds like, pretty much. Can add an avatar or name, group names can be null and default (through function calls, to GroupChat.GetGroupName(), be careful) to the user's display name (username if null) + Group Chat.
  * Checklist for which friends you want to invite upon creation

## Changes
* Redesigned the friends component; added more functionality for interactivity and usability
  * Numbers are displayed in tab headers where it is relevant to do so
  * Redesigned the 'find friends' component, and extracted to pop-up. 
  * Redesigned the friends list, block list, added a group invites list, etc.

Honestly there's probably a lot more I just cannot be asked/remember everything, but visually I'd say we're pretty much all set for the interface

### CODE REVIEW

File: App/Components/Interaction/CompFriendList.razor [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    private List<User> Friends { get; set; } = [];
+
+    protected override async Task OnInitializedAsync()
+    {
+        Self = (await AuthService.GetUser())!;
+        UserInteractionService.AddSubscriber(this);
+        await GetFriendData();
+    }
+
+    public RenderFragment GetFriendFragment(User user)
```
Comment (Bawnorton) [inline App/Components/Interaction/CompFriendList.razor:60]: I redesigned this in my branch to look like:

https://github.com/Bawnorton/Bamboozlers/blob/19beb6d771603bf7924ae6edffa722b2a60dfc8c/App/Components/Friend/CurrentFriends.razor#L45-L56

Which I believe looks better. The Message friend button being readily available is also a good idea.

File: App/Components/MainVisual/CompMainView.razor [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+                        @if (Blocked.Count > 0)
+                        {
+                            <Badge ElementId="BlockCountBadge"
+                                   Color="Color.Warning">
+                                @Blocked.Count
+                            </Badge>
+                        }
+                        Blocked
+                    </Tab>
+                    <CompActionButton ColumnWrapped="@false"
```
Comment (Bawnorton) [inline App/Components/MainVisual/CompMainView.razor:66]: Wrap in
```
<Div class="p-1">
   ...
</Div>
```

**Review by Bawnorton — CHANGES_REQUESTED** [review]
Looks great so far, couple styling recommendations from my own pr.

`UserSettingsTest_CompDeleteAccount` is failing:
```
Assert.DoesNotContain() Failure: Sub-string found
                         ↓ (pos 16)
String: "tab-pane active show"
Found:  "show"
   at Tests.UserTests.UserSettingsTests.CheckElementClassContains(IElement e, String whichClass, Boolean notContains) in /Users/bawnorton/Documents/Developer/C#/Bamboozlers/Tests/UserTests/UserSettingsTests.cs:line 898
   ```
   Everything else runs as expected.

File: App/Components/Interaction/CompFriendList.razor (hunk 1 above)
Comment (Bawnorton) [reply App/Components/Interaction/CompFriendList.razor:60]: https://github.com/Bawnorton/Bamboozlers/assets/18416784/f899983a-e127-48b9-9f55-3a1284d2e12f

Comment (bhunt02) [reply App/Components/Interaction/CompFriendList.razor:60]: will swap what i've got out for this

File: App/Components/Interaction/CompAddFriend.razor [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                          MaskType="MaskType.RegEx"
+                          EditMask="^(?:_?[a-zA-Z0-9]+)*_?$"
+                          Placeholder="Find users by their username or display name"/>
+                <div style="display: inline-flex">
+                    <CompActionButton ActionCallback="@(async () => await PerformSearch())"
+                                      IconName="@IconName.Search"
+                                      ColumnWrapped="@false"/>
+                </div>
+            </div>
+            <div style="display: block; margin-left: 5px; margin-right: 5px;" class="settings-container scrollbar-container">
```
Comment (Bawnorton) [inline App/Components/Interaction/CompAddFriend.razor:33]: `settings-container` has `padding 0 10%` which looks out of place, full-width looks better consider modals already have some padding.

Also to not, `display: block` is default for divs and doesn't need to be specified

File: App/Components/Interaction/CompAddFriend.razor [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+<ModalContent Centered Size="ModalSize.Large">
+    <ModalHeader>
+        <ModalTitle>Add Friends</ModalTitle>
+    </ModalHeader>
+    <ModalBody>
+        <Heading>Find users by their username or display name.</Heading>
+        <Divider/>
+        <Div Style="display: block">
+            <div style="display: block;">
+                <TextEdit @bind-Text="SearchQuery"
```
Comment (Bawnorton) [inline App/Components/Interaction/CompAddFriend.razor:21]: Pressing enter should search, my branch has a keypress service listener thingy if you want to defer that to post merge

**Review by Bawnorton — CHANGES_REQUESTED** [review]

**Review by Bawnorton — CHANGES_REQUESTED** [review]
Looks good, `NavLayoutTests_FindAndOpenGroups` fails:
```
Assert.Equal() Failure: Strings differ
           ↓ (pos 0)
Expected: "TestGroupChat"
Actual:   "Hello, TestUser0!"
           ↑ (pos 0)
   at Tests.CoreTests.NavLayoutTests.NavLayoutTests_FindAndOpenGroups() in /Users/bawnorton/Documents/Developer/C#/Bamboozlers/Tests/CoreTests/NavLayoutTests.cs:line 75
   ```
   `NavLayoutTests_CheckReactionToInvalidChat` Fails:
```
Assert.Equal() Failure: Strings differ
           ↓ (pos 0)
Expected: "TestUser1"
Actual:   "Hello, TestUser0!"
           ↑ (pos 0)
   at Tests.CoreTests.NavLayoutTests.NavLayoutTests_CheckReactionToInvalidChat() in /Users/bawnorton/Documents/Developer/C#/Bamboozlers/Tests/CoreTests/NavLayoutTests.cs:line 166
   ```

**Review by Bawnorton — APPROVED** [review]

**Pull request merged**